File: 백준 문제 풀이 모음/class 4/16953.py
import sys

input = sys.stdin.readline
from collections import deque

a, b = map(int, input().split())


def dfs(x) :

    queue = deque()
    queue.append([x])

    
    count = 0
    state = False
    while queue :
    
        node = queue.popleft()

        if len(node) == 0 :
            print(-1)
            break    
        if state :
            print(count+1)
            break

        tmp = []
        for i in node :
            if i* 2 < b :
                tmp.append(i * 2)
            elif i * 2 == b :
                tmp.append(i * 2)
                state = True

            if i* 10 + 1 <b :
                tmp.append(i * 10 + 1)
            elif i * 10 + 1 == b :
                tmp.append(i * 10 + 1)
                state = True
        queue.append(tmp)
        count += 1

dfs(a)

File: 백준 문제 풀이 모음/kotlin/class 4/16953.kt
import java.util.PriorityQueue

fun main() = with(System.`in`.bufferedReader()){
    val (a, b) = readLine().split(' ').toList().map { it.toLong() }

    bfs(a, b)


}

fun bfs(a: Long, b: Long) {

    val deque = ArrayDeque<MutableSet<Long>>()
    deque.add(mutableSetOf(a))
    var count = 1

    while (true) {

        if (deque[0].isEmpty()) {
            println(-1)
            break
        }

        val node = deque.removeFirst() // (2)

        if (node.contains(b)) {
            println(count)
            break
        }

        val new_set = mutableSetOf<Long>()
        node.forEach {
            if (it * 2 <= b) {
                new_set.add(it * 2)
            }

            if (it * 10 + 1 <= b) {
                new_set.add(it * 10 + 1)
            }
        }

        deque.add(new_set)
        count ++
    }
}
